keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+* An optional data source that can either be a local snapshot of the application or a backup in a remote location, from where the application can be restored or cloned.
+* The name of an optional `StatefulApplicationClass` that defines common parameters for a class of `StatefulApplication`s. The parameters are used by the handler when the application is provisioned, e.g., cloned or restored from a data source. See details about `StatefulApplicationClass` below.
+* An optional list of `ValueSubstitutionRule`s that define how to substitute old values with new values for certain types of values in the application's Kubernetes resource definitions, when the application is provisioned.
+
+The following data management operations can be performed on a `StatefulApplication`:
+
+* Taking a local snapshot of the application and store the snapshot locally within the cluster.
+* Taking a backup of the application to a remote location, e.g., a Google Cloud Storage (GCS) or S3 bucket.
+* Restoring the application from either a local snapshot within the cluster or a backup in a remote location.
+* Cloning the application from either a local snapshot within the cluster or a backup in a remote location.
+* Rolling back the application state to that captured in either a local snapshot within the cluster or a backup in a remote location.
+
+What operation gets triggered for a `StatefulApplication` depends on what values it contains in the two fields above, as the following table summarizes.
+
+| | Has Data Source | Has no Data Source |
+| ------ | ------ | ------ |
+| References an Application | No action needed; it's just a wrapper around an `Application` resource. | No action needed; this is not a meaningful combination. |
+| References no Application | Creates (clones or restores) the application from the given data source. | No action needed; this state is the result of creating an application from the given data source. |
+
+The detailed definition of `StatefulApplication` is as follows.
+
+```go
+type StatefulApplicationSpec struct {
+  // ApplicationClassName is the name of the StatefulApplicationClass to use creating the application,
+	// e.g., restoring the application from a data source. If this is not specified, the default one of
+	// the cluster is used.
+	// +optional
+  ApplicationClassName *string `json:"applicationClassName,omitempty"`
+  
+  // ApplicationName is the name of the Application (https://github.com/kubernetes-sigs/application)
+  // object, which represents an application and gives the list of Kubernetes resources that make up the
+  // application in its status. When doing an application restoration, this is the name of the Application
+  // resource to be created if set. The name of this StatefulApplication resource will be used if this is
+  // not set. It’s assumed that the Application resource is in the same namespace of it exists.
+  // +optional
+  ApplicationName *string `json:"applicationName,omitempty"`
```

Comment:
Will there be any validation on whether the application specified is really a stateful application or not?  Will the validation done by the controller or can it be handled in API validation?

Reply:
No. There's no validation on this. I'm thinking that the same mechanism could also be used for taking snapshots of resource definitions of applications that do not use PVCs.